prov/efa/src/rdm/efa_rdm_pke_utils.c

```diff
@@ -82,13 +82,17 @@ ssize_t efa_rdm_pke_init_payload_from_ope(struct efa_rdm_pke *pke,
 	}
 
 	/*
-	 * Copy can be avoid if the following 2 conditions are true:
+	 * Copy can be avoid if the following 3 conditions are true:
 	 * 1. EFA can directly access the memory
-	 * 2. data to be send is in 1 iov, because device only support 2 iov, and we use
-	 *    1st iov for header.
+	 * 2. data to be send is in 1 iov, because device only support 2 iov,
+	 * and we use 1st iov for header.
+	 * 3. FI_INJECT transfer flag is not set. FI_INJECT requires outbound
+	 * buffer to be reusable immediately after function returns, so without
+	 * the inline HW support, a copy is needed.
 	 */
 	if (efa_can_access_memory &&
-	    (tx_iov_offset + data_size <= ope->iov[tx_iov_index].iov_len)) {
+	    (tx_iov_offset + data_size <= ope->iov[tx_iov_index].iov_len) &&
+	    !(ope->fi_flags & FI_INJECT)) {
```

Comment:
I don't think this is correct.

FI_INJECT can still avoid copy when the buffer is on host and size is within inline. It seems we need some more refactor here.